simulation/simulation.go

```diff
@@ -114,27 +110,41 @@ func (e *entitySet) create(ctx context.Context, sim *Simulation) {
 		sim.error(errors.NotFoundf("entity %q", e.Entity))
 		return
 	}
+	e.cfg = cfg
+
 	// determine how many entities are to be created
 	c := cardinality(e.Cardinality)
 	numberOfEntities, err := c.Value(e.attributes)
 	if err != nil {
 		sim.error(errors.Trace(err))
 		return
 	}
-	// create the timer that determines the cadence of entity
-	// creation
-	timer := newTimer(e.Timer)
-	for i := 0; i < numberOfEntities; i++ {
-		err = timer.Next(ctx)
-		if err != nil {
-			sim.error(errors.Trace(err))
-			return
-		}
-		createEntity(ctx, cfg, copyAttributes(e.attributes), sim)
-	}
+	e.numberOfEntities = numberOfEntities
+	e.numberOfCreatedEntities = 0
+
+	// add a job that will all all the entities
+	sim.e.addJob(ctx, newDelay(e.Timer), func() {
+		e.createEntity(ctx, sim)
+	})
+
 	return
 }
 
+func (e *entitySet) createEntity(ctx context.Context, sim *Simulation) {
+	e.mu.Lock()
+	if e.numberOfCreatedEntities >= e.numberOfEntities {
+		return
```

Comment:
Lock would still be held here. I think you'd want to move the defer unlock up, to immediately follow the lock.

Reply:
fixed